uat/custom-components/src/main/java/com/aws/greengrass/utils/JsonMerger.java

```diff
@@ -0,0 +1,213 @@
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+public class JsonMerger {
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    private JsonMerger() {
+    }
+
+    public static void merge(final JsonNode source, final JsonNode patch) {
+
+        //ValueNodes are immutable, thus we cannot merge a new value into it.
+        if (source.isValueNode()) {
+            throw new IllegalStateException("Cannot merge update a value node.  ValueNodes are immutable");
+        }
+
+        //if both nodes are objects then do a recursive patch
+        if (source.isObject() && patch.isObject()) {
+            merge((ObjectNode) source, (ObjectNode) patch);
+            return;
+        }
+
+        //if both nodes are arrays then overwrite the source array's contents
+        if (source.isArray() && patch.isArray()) {
+            ((ArrayNode) source).removeAll();
+            ((ArrayNode) source).addAll((ArrayNode) patch);
+            return;
+        }
+
+        throw new IllegalArgumentException("Merge only works with Json objects whose underlying node are the same type.");
+
+    }
+
+    private static void merge(final ObjectNode source, final ObjectNode patch) {
+
+        final Iterator<String> fieldIter = patch.fieldNames();
+        while (fieldIter.hasNext()) {
+            final String field = fieldIter.next();
+
+            final JsonNode originalValue = source.get(field);
+            final JsonNode patchValue = patch.get(field);
+
+            //if the patch value is set to null then wipe out the value from original if present
+            if (isNull(patchValue)) {
+                if (!isNull(originalValue)) {
+                    source.remove(field);
+                }
+                continue;
+            }
+
+            //if the source doesn't have the value then add it
+            if (isNull(originalValue)) {
+                if (patchValue.isObject()) {
+                    final ObjectNode child = createMergeTree((ObjectNode) patchValue);
+                    if (child != null) {
+                        source.set(field, child);
+                    }
+                } else {
+                    //TODO this will allow arrays to leak wtih null values in them.  We need
+                    //to re-think how to handle arrays
+                    source.set(field, patchValue);
+                }
+                continue;
+            }
+
+            //if we got objects then recurse
+            if (originalValue.isObject() && patchValue.isObject()) {
+                merge((ObjectNode) originalValue, (ObjectNode) patchValue);
+                continue;
+            }
+
+            //If sourceValue and patchValue are not objects then we just update the source field
+            //to point the new patch value.
+            source.set(field, patchValue);
+        }
+    }
+
+    static void merge(final JsonNode source, final JsonPointer pointer, final JsonNode patch) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 98-120, lines 123-137, lines 140-164 into separate methods.